File: joomla/ingest.py
#!/usr/bin/env python3
"""
Joomla Extension Directory ingester.

NOTE: JED does not centrally host downloads. Each extension links to the developer's
own server. This script:
  1. Queries the JED API for extension listings
  2. Extracts GitHub repo URLs from the metadata (most popular extensions are on GitHub)
  3. Downloads the latest GitHub release ZIP for those
  4. Logs the rest to output/joomla/manual_review.tsv for manual follow-up

Usage:
    python -m joomla.ingest --max-extensions 100 --output-dir output/
    python -m joomla.ingest --max-extensions 100 --github-token ghp_xxx --output-dir output/
"""
import argparse
import logging
import re
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from common.state import DownloadState
from common.http import RateLimitedSession

ECOSYSTEM = "joomla"
JED_API = "https://extensions.joomla.org/index.php"
GITHUB_API = "https://api.github.com"
PAGE_SIZE = 20

logger = logging.getLogger(__name__)


def list_extensions(session: RateLimitedSession, start: int) -> list:
    """Fetch one page of extensions from JED. Returns empty list on any failure."""
    params = {
        "option": "com_jed",
        "view": "extension",
        "format": "json",
        "limit": PAGE_SIZE,
        "start": start,
        "orderby": "popular",
    }
    try:
        resp = session.get(JED_API, params=params)
        data = resp.json()
        # JED API response format has changed over time:
        #   - bare list (current as of 2026)
        #   - {"data": [...]}
        #   - {"extensions": [...]}
        #   - {"list": [...]}
        if isinstance(data, list):
            return data
        return data.get("data", data.get("extensions", data.get("list", [])))
    except Exception as exc:
        logger.warning("JED page start=%d failed: %s", start, exc)
        return []


def _jed_str(field) -> str:
    """Unwrap a JED field that may be a plain string or a {value, text} dict."""
    if isinstance(field, dict):
        return field.get("value", "") or field.get("text", "")
    return str(field) if field else ""


def extract_github_repo(extension: dict) -> str | None:
    """Extract 'owner/repo' from any URL-like field in the extension record."""
    text = " ".join(_jed_str(extension.get(k, "")) for k in [
        "url", "download_url", "repository_url", "homepage",
        "homepage_link", "download_link", "download_integration_url",
    ])
    m = re.search(r"github\.com[:/]([^/\s]+/[^/\s.#?]+)", text)
    if m:
        repo = m.group(1).rstrip("/")
        if repo.endswith(".git"):
            repo = repo[:-4]
        return repo
    return None


def download_from_github(session: RateLimitedSession, state: DownloadState,
                          github_token: str | None, repo: str,
                          slug: str, output_dir: str) -> bool:
    """Download latest release ZIP from a GitHub repo. Returns True if downloaded."""
    headers = {}
    if github_token:
        headers["Authorization"] = f"token {github_token}"
    try:
        resp = session.get(f"{GITHUB_API}/repos/{repo}/releases/latest", headers=headers)
        release = resp.json()
        version = release.get("tag_name", "unknown")
        zipball_url = release.get("zipball_url")
        if not zipball_url:
            logger.debug("No zipball_url for %s", repo)
            return False
        if state.is_downloaded(ECOSYSTEM, slug, version):
            logger.debug("Skip %s %s", slug, version)
            return False
        dest = str(Path(output_dir) / ECOSYSTEM / f"{slug}.{version}.zip")
        file_hash = session.download(zipball_url, dest)
        state.record(ECOSYSTEM, slug, version, dest, file_hash)
        logger.info("Downloaded %s %s (GitHub:%s) -> %s", slug, version, repo, dest)
        return True
    except Exception as exc:
        logger.warning("GitHub download failed for %s (%s): %s", slug, repo, exc)
        return False


def main():
    parser = argparse.ArgumentParser(description="Ingest Joomla extensions")
    parser.add_argument("--max-extensions", type=int, default=0,
                        help="Stop after N downloads (0 = unlimited)")
    parser.add_argument("--output-dir", default="output")
    parser.add_argument("--db", default="data/downloads.db")
    parser.add_argument("--github-token", default=None,
                        help="GitHub PAT for higher API rate limits (60 -> 5000 req/hr)")
    parser.add_argument("--requests-per-second", type=float, default=1.0)
    parser.add_argument("-v", "--verbose", action="store_true")
    args = parser.parse_args()

    logging.basicConfig(
        level=logging.DEBUG if args.verbose else logging.INFO,
        format="%(asctime)s %(levelname)s %(message)s",
    )

    session = RateLimitedSession(args.requests_per_second)
    state = DownloadState(args.db)
    manual_review = []
    downloaded = 0
    start = 0

    while True:
        extensions = list_extensions(session, start)
        if not extensions:
            logger.info("No extensions returned at start=%d, stopping.", start)
            break

        for ext in extensions:
            if args.max_extensions and downloaded >= args.max_extensions:
                logger.info("Reached --max-extensions=%d, stopping.", args.max_extensions)
                break

            # core_alias is the canonical slug in the new JED format
            slug = str(
                _jed_str(ext.get("core_alias"))
                or ext.get("alias") or ext.get("element") or ext.get("id")
                or "unknown"
            )
            github_repo = extract_github_repo(ext)

            if github_repo:
                if download_from_github(
                    session, state, args.github_token, github_repo, slug, args.output_dir
                ):
                    downloaded += 1
            else:
                fallback_url = (
                    _jed_str(ext.get("download_link"))
                    or _jed_str(ext.get("homepage_link"))
                    or ext.get("url") or ext.get("download_url", "")
                )
                manual_review.append(f"{slug}\t{fallback_url}")
                logger.debug("No GitHub URL for %s; logged for manual review", slug)
        else:
            start += PAGE_SIZE
            continue
        break  # inner for-loop hit break (max reached)

    if manual_review:
        review_path = Path(args.output_dir) / ECOSYSTEM / "manual_review.tsv"
        review_path.parent.mkdir(parents=True, exist_ok=True)
        with open(review_path, "w") as f:
            f.write("slug\turl\n")
            f.write("\n".join(manual_review))
        logger.info(
            "Logged %d extensions for manual review -> %s",
            len(manual_review), review_path,
        )

    logger.info("Done. %d downloaded this run. DB totals: %s", downloaded, state.stats())


if __name__ == "__main__":
    main()

File: ts/src/joomla/ingest.ts
/**
 * Joomla Extension Directory ingester.
 * JED does not host downloads centrally. GitHub-hosted extensions are downloaded
 * via GitHub Releases. Others are logged to manual_review.tsv.
 *
 * Usage:
 *   npx tsx src/joomla/ingest.ts --max-extensions 100
 *   npx tsx src/joomla/ingest.ts --max-extensions 100 --github-token ghp_xxx
 */
import { join } from "path";
import { mkdirSync, createWriteStream } from "fs";
import { program } from "commander";
import { RateLimitedClient } from "../common/http.js";
import { DownloadState } from "../common/state.js";
import type { PackageMetadata } from "../common/types.js";

export const ECOSYSTEM = "joomla";
const JED_API = "https://extensions.joomla.org/index.php";
const GITHUB_API = "https://api.github.com";
const PAGE_SIZE = 20;

// JED field values come as either plain strings or {value, text} objects
type JedField = string | { value?: string; text?: string } | null | undefined;

function jedStr(field: JedField): string {
  if (!field) return "";
  if (typeof field === "string") return field;
  return field.value ?? field.text ?? "";
}

function jedNum(field: JedField): number | undefined {
  const s = jedStr(field);
  const n = parseFloat(s);
  return isNaN(n) ? undefined : n;
}

interface JedExtension {
  id?: string | number;
  core_title?: JedField;
  core_alias?: JedField;
  core_body?: JedField;
  core_created_time?: JedField;
  core_modified_time?: JedField;
  core_hits?: JedField;
  version?: JedField;
  type?: JedField;
  license?: JedField;
  score?: JedField;
  num_reviews?: JedField;
  functionality?: JedField;
  ease_of_use?: JedField;
  support?: JedField;
  documentation?: JedField;
  value_for_money?: JedField;
  download_link?: JedField;
  homepage_link?: JedField;
  download_integration_url?: JedField;
  approved?: JedField;
  approved_time?: JedField;
  community_choice?: JedField;
  popular?: JedField;
  tags?: unknown;
  // old format
  url?: string;
  download_url?: string;
  alias?: string;
  element?: string;
}

export function listExtensions(
  client: RateLimitedClient,
  start: number
): Promise<JedExtension[]> {
  const params = new URLSearchParams({
    option: "com_jed",
    view: "extension",
    format: "json",
    limit: String(PAGE_SIZE),
    start: String(start),
    orderby: "popular",
  });
  return client
    .getJson<JedExtension[] | { data?: JedExtension[]; extensions?: JedExtension[]; list?: JedExtension[] }>(
      `${JED_API}?${params}`
    )
    .then((data) => {
      if (Array.isArray(data)) return data;
      return (
        (data as { data?: JedExtension[] }).data ??
        (data as { extensions?: JedExtension[] }).extensions ??
        (data as { list?: JedExtension[] }).list ??
        []
      );
    })
    .catch(() => []);
}

export function extractGithubRepo(ext: JedExtension): string | null {
  const text = [
    ext.url ?? "",
    ext.download_url ?? "",
    jedStr(ext.homepage_link),
    jedStr(ext.download_link),
    jedStr(ext.download_integration_url),
  ].join(" ");

  const match = text.match(/github\.com[:/]([^/\s]+\/[^/\s.#?]+)/);
  if (!match) return null;
  return match[1].replace(/\/$/, "").replace(/\.git$/, "");
}

export function toMetadata(ext: JedExtension): PackageMetadata {
  const slug =
    jedStr(ext.core_alias) ||
    (ext.alias as string) ||
    (ext.element as string) ||
    String(ext.id ?? "unknown");

  const tags: string[] = [];
  if (ext.tags && typeof ext.tags === "object") {
    for (const tag of Object.values(ext.tags as Record<string, unknown>)) {
      if (typeof tag === "string") tags.push(tag);
      else if (tag && typeof tag === "object" && "name" in tag) {
        tags.push(String((tag as { name: string }).name));
      }
    }
  }

  return {
    ecosystem: ECOSYSTEM,
    slug,
    name: jedStr(ext.core_title) || slug,
    currentVersion: jedStr(ext.version) || undefined,
    description: jedStr(ext.core_body)?.slice(0, 500) || undefined,
    homepageUrl: jedStr(ext.homepage_link) || undefined,
    downloadUrl: jedStr(ext.download_link) || jedStr(ext.download_integration_url) || undefined,
    license: jedStr(ext.license) || undefined,
    tags: tags.length > 0 ? tags : undefined,
    registryAddedDate: jedStr(ext.core_created_time) || undefined,
    lastUpdatedDate: jedStr(ext.core_modified_time) || undefined,
    downloadCount: jedNum(ext.core_hits),  // page hits, best available proxy
    rating: jedNum(ext.score),
    numRatings: jedNum(ext.num_reviews),
    rawMetadata: JSON.stringify({
      type: jedStr(ext.type),
      approved_time: jedStr(ext.approved_time),
      community_choice: jedStr(ext.community_choice),
      popular: jedStr(ext.popular),
      review_scores: {
        functionality: jedNum(ext.functionality),
        ease_of_use: jedNum(ext.ease_of_use),
        support: jedNum(ext.support),
        documentation: jedNum(ext.documentation),
        value_for_money: jedNum(ext.value_for_money),
      },
    }),
  };
}

interface GitHubRelease {
  tag_name?: string;
  zipball_url?: string;
}

export async function downloadFromGithub(
  client: RateLimitedClient,
  state: DownloadState,
  githubToken: string | undefined,
  repo: string,
  slug: string,
  outputDir: string
): Promise<boolean> {
  const headers: Record<string, string> = {};
  if (githubToken) headers["Authorization"] = `token ${githubToken}`;

  try {
    const release = await client.getJson<GitHubRelease>(
      `${GITHUB_API}/repos/${repo}/releases/latest`,
      { headers }
    );
    const version = release.tag_name ?? "unknown";
    const zipballUrl = release.zipball_url;
    if (!zipballUrl) return false;

    if (state.isDownloaded(ECOSYSTEM, slug, version)) return false;

    const destPath = join(outputDir, ECOSYSTEM, `${slug}.${version}.zip`);
    const fileHash = await client.download(zipballUrl, destPath);
    state.record(ECOSYSTEM, slug, version, destPath, fileHash);
    return true;
  } catch (err) {
    console.warn(`[joomla] GitHub download failed for ${repo}: ${(err as Error).message}`);
    return false;
  }
}

async function main() {
  program
    .option("--max-extensions <n>", "stop after N downloads (0 = unlimited)", "0")
    .option("--output-dir <dir>", "output directory", "output")
    .option("--db <path>", "sqlite db path", "data/downloads.db")
    .option("--github-token <token>", "GitHub PAT for higher rate limits")
    .option("--requests-per-second <n>", "rate limit", "1")
    .option("-v, --verbose", "verbose logging")
    .parse();

  const opts = program.opts();
  const maxExtensions = parseInt(opts.maxExtensions);

  const client = new RateLimitedClient(parseFloat(opts.requestsPerSecond));
  const state = new DownloadState(opts.db);
  const seenSlugs = new Set<string>();
  const manualReview: string[] = [];

  let downloaded = 0;
  let start = 0;
  let stop = false;

  while (!stop) {
    const extensions = await listExtensions(client, start);
    if (extensions.length === 0) {
      console.log(`[joomla] No extensions at start=${start}, stopping.`);
      break;
    }

    for (const ext of extensions) {
      const meta = toMetadata(ext);
      seenSlugs.add(meta.slug);
      state.upsertMetadata(meta);

      if (maxExtensions > 0 && downloaded >= maxExtensions) {
        console.log(`[joomla] Reached --max-extensions=${maxExtensions}, stopping.`);
        stop = true;
        break;
      }

      const repo = extractGithubRepo(ext);
      if (repo) {
        if (
          await downloadFromGithub(
            client,
            state,
            opts.githubToken,
            repo,
            meta.slug,
            opts.outputDir
          )
        ) {
          downloaded++;
          console.log(`[joomla] Downloaded ${meta.slug} (GitHub:${repo})`);
        }
      } else {
        const fallbackUrl = meta.downloadUrl ?? "";
        manualReview.push(`${meta.slug}\t${fallbackUrl}`);
      }
    }

    if (!stop) start += PAGE_SIZE;
  }

  if (manualReview.length > 0) {
    const reviewDir = join(opts.outputDir, ECOSYSTEM);
    mkdirSync(reviewDir, { recursive: true });
    const reviewPath = join(reviewDir, "manual_review.tsv");
    const stream = createWriteStream(reviewPath);
    stream.write("slug\turl\n");
    stream.write(manualReview.join("\n"));
    stream.end();
    console.log(`[joomla] Logged ${manualReview.length} extensions for manual review -> ${reviewPath}`);
  }

  const removed = state.markRemovedIfAbsent(ECOSYSTEM, seenSlugs);
  if (removed > 0) console.log(`[joomla] Marked ${removed} extensions as removed.`);
  console.log(`[joomla] Done. ${downloaded} downloaded. DB totals:`, state.stats());
  console.log(`[joomla] Metadata tracked:`, state.metadataStats());
  state.close();
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
